src/react/GroupsPage.tsx

```diff
@@ -19,33 +19,51 @@
 import React, { useEffect, useState } from 'react';
 import { Link } from 'react-router-dom';
 import { Button, Spinner, Table } from 'react-bootstrap';
+import AstarteClient, { AstarteDevice } from 'astarte-client';
 
 import SingleCardPage from './ui/SingleCardPage';
 import { useAlerts } from './AlertManager';
 
-export default ({ astarte, history }) => {
-  const [phase, setPhase] = useState('loading');
-  const [groups, setGroups] = useState(null);
+interface Props {
+  astarte: AstarteClient;
+  history: any;
+}
+
+interface GroupState {
+  name: string;
+  loading: boolean;
+  totalDevices: number;
+  connectedDevices: number;
+}
+
+type GroupMap = Map<string, GroupState>;
+
+export default ({ astarte, history }: Props): React.ReactElement => {
+  const [phase, setPhase] = useState<'ok' | 'loading' | 'err'>('loading');
+  const [groups, setGroups] = useState<GroupMap | null>(null);
   const pageAlerts = useAlerts();
 
   useEffect(() => {
-    const handleDeviceList = (groupName, devices) => {
-      setGroups((groupMap) => {
-        const newGroupState = groupMap.get(groupName);
+    const handleDeviceList = (groupName: string, devices: AstarteDevice[]) => {
+      setGroups((groupMap: GroupMap | null) => {
+        const newGroupMap = groupMap ? new Map(groupMap) : new Map();
+        const newGroupState = newGroupMap.get(groupName);
         newGroupState.loading = false;
         newGroupState.totalDevices = devices.length;
         const connectedDevices = devices.filter((device) => device.isConnected);
         newGroupState.connectedDevices = connectedDevices.length;
-        groupMap.set(groupName, newGroupState);
-        return new Map(groupMap);
+        newGroupMap.set(groupName, newGroupState);
+        return newGroupMap;
       });
     };
-    const handleGroupsRequest = (groupNames) => {
-      const groupMap = groupNames.reduce((acc, groupName) => {
+    const handleGroupsRequest = (groupNames: string[]) => {
+      const groupMap: GroupMap = groupNames.reduce((acc, groupName) => {
         acc.set(groupName, { name: groupName, loading: true });
         return acc;
       }, new Map());
-      groupMap.forEach((value, groupName) => {
+      setGroups(groupMap);
+      setPhase('ok');
+      groupMap.forEach((_, groupName) => {
```

Comment:
if you dont' use the first argument, you can just skip it `(, groupName)` assigning it to `_` may conflict with lodash

Reply:
Typescript will yield a parsing error for `(, groupName)`. I also think you could mistakenly read it as 1 argument, not noticing the comma.
I left `(groupState, groupName)` for the sake of readability.